### PR TITLE
[#169683271] Write Script that Flips an review environment to .ninja or .com

### PR DESCRIPTION
**Pull request by dereksweet**
[PT-#169683271](https://www.pivotaltracker.com/story/show/169683271)

## Description
This PR adds a new job to our `heroku-review-envs` scripts. 

### heroku-config-var-set
This job allows you to set any number of config variables on a specific heroku application. When calling it you give it the Heroku API token as an environment variable, then pass in a number of arguments specifying the heroku app you want to set it them on and the config variables you want to set. 

I followed the same pattern that was present in the `APP_REF` for specifying config variables since passing raw JSON did not work well with github workflows. You must specify your CONFIG_VARS like so: 

`CONFIG_VARS=var1%value1|var2%value2|var3%value3`

`%` is used in place of `=`
`|` is used to separate variables

Usage: 
**build**
`docker build -t config-var-set heroku-config-var-set/`

**run**
`docker run -e HEROKU_API_TOKEN={Heroku Token} config-var-set PP_ORIGIN=web APP_TARGET=web APP_PREFIX=trr CONFIG_VARS=OKTA_AUTH_SERVER_BASE_URL%https://login.therealreal.com/oauth2/default/|OKTA_CLIENT_ID%0oa18r3unfqJuF9Y4357|OKTA_CLIENT_SECRET%{Client Secret}`

## Motivation and Context
We want to be able to easily flip between .com and .ninja Okta settings for review environments. These jobs will facilitate that by adding a label to the PR that controls the arguments that are passed into the above jobs. 

## Testing
- For testing purposes you can use the following PR: https://github.com/TheRealReal/real-server/pull/16711
- If you add the `okta-ninja` label, or remove it, it should adjust the Okta envrionment that the review env is configured for
- Make sure all checks are finished running before you attempt to test the adjusted settings
- Visit the following URL with a brand new incognito browser for each test: https://trr-web-pr-16711.herokuapp.com/admin/okta
- If you've configured it for .com, you should be able to use your own .com account to login
- If you've configured it for .ninja, use one of the test users like `price.administrator@therealreal.ninja`
- Nothing should be tested much further than login, that is all that is affected by this. The Okta role groups should work as expected, however, so price.administrator should have price administration privileges, for example. 

## Type of Changes

- [ ] Bug fix _(a non-breaking change which fixes an issue)_
- [x] New feature _(a non-breaking change which adds functionality)_
- [ ] Breaking change _(fix or feature that changes existing functionality)_
- [ ] Bump dependencies
- [ ] Automated Testing / Missing tests
- [ ] Documentation

## Checklist:

- [ ] Yes, I have added tests to cover my changes.
- [x] No, We don't have a test suite in this project yet.
- [ ] My change requires a change to the documentation.
- [ ] My change requires release strategy _(uncomment the Release strategy below)_

### CODE REVIEW

**Comment by dereksweet** [comment]
> if `okta-whitelist-toggle` is essentially just a `destroy` and `create`, it doesn't feel very DRY. Can we see how much of a difference it makes in runtime before adding this?

I was going to remove the `create` task following this deploy and once the workflows were adjusted as it wouldn't be needed anymore. There will be a very negligible difference in runtime, isn't worth doing for that. If you'd rather I make it two separate create and destroy steps I can do that instead and remove the toggle task completely.

**Comment by trr-aaronchu** [comment]
In that case, maybe we should modify the `create` action to do a `delete-if-different-then-create` instead? This should be backward compatible and offer a simple rollout?

**Comment by dereksweet** [comment]
> In that case, maybe we should modify the `create` action to do a `delete-if-different-then-create` instead? This should be backward compatible and offer a simple rollout?

That's pretty much what the new toggle task does. It is basically a new create that will remove it from the other system if it's there, and do nothing if it's not. Or maybe I'm not understanding you correctly?

**Comment by trr-aaronchu** [comment]
What I mean is that if we had a PR that simply modified the `create` action instead of introducing a new one, that might be an easier migration path.

**Comment by dereksweet** [comment]
> What I mean is that if we had a PR that simply modified the `create` action instead of introducing a new one, that might be an easier migration path.

OK I'll see if I can do that without changing the syntax of the github workflow, but I don't think that's going to be possible since it will need the new arguments for communicating with .ninja anyway. I'm 99% sure the workflows would have to be updated regardless so I felt this was the safest way to do things so the existing workflows would continue to work after this is deployed since these are two brand new tasks.

**Comment by dereksweet** [comment]
@trr-aaronchu it will not be possible to keep it backwards compatible if I change the existing create script as it will have to expect new arguments being passed in. I can, however, make it work with no changes to the create and destroy tasks by replacing the two toggle steps in the new workflow to 4 new create and delete steps. Then this PR could just remove the toggle task completely. Would you prefer we do that?

**Comment by dereksweet** [comment]
@trr-aaronchu for your reference here is what the steps would look like for the Toggle step that is in this PR (will compare to the 4 create/delete steps after): 

```
- name: Flips to Okta .com whitelist
  uses: TheRealReal/heroku-review-envs/okta-whitelist-toggle@master
  if: "!contains( toJSON(github.event.pull_request.labels), 'okta-ninja')"
  env:
    OKTA_API_TOKEN: ${{ secrets.OKTA_API_TOKEN }}
    OKTA_API_URL: https://login.therealreal.com/oauth2/v1/clients/0oa18r3unfqJuF9Y4357
    OKTA_API_TOKEN_ALT: ${{ secrets.OKTA_API_TOKEN_NINJA }}
    OKTA_API_URL_ALT: https://login.therealreal.ninja/oauth2/v1/clients/0oalprgcvn7yBQMUU0h7
  with:
    args: OKTA_TARGET=main APP_ORIGIN=web APP_TARGET=web APP_PREFIX=trr URL_TARGET=https://%s.herokuapp.com/admin/okta
- name: Flips to Okta .ninja whitelist
  uses: TheRealReal/heroku-review-envs/okta-whitelist-toggle@master
  if: "contains( toJSON(github.event.pull_request.labels), 'okta-ninja')"
  env:
    OKTA_API_TOKEN: ${{ secrets.OKTA_API_TOKEN }}
    OKTA_API_URL: https://login.therealreal.com/oauth2/v1/clients/0oa18r3unfqJuF9Y4357
    OKTA_API_TOKEN_ALT: ${{ secrets.OKTA_API_TOKEN_NINJA }}
    OKTA_API_URL_ALT: https://login.therealreal.ninja/oauth2/v1/clients/0oalprgcvn7yBQMUU0h7
  with:
    args: OKTA_TARGET=alt APP_ORIGIN=web APP_TARGET=web APP_PREFIX=trr URL_TARGET=https://%s.herokuapp.com/admin/okta
```

Here is what it would look like with the 4 create/delete steps: 

```
- name: Add To Okta Whitelist (com)
  uses: TheRealReal/heroku-review-envs/okta-whitelist-url-create@master
  if: "!contains( toJSON(github.event.pull_request.labels), 'okta-ninja')"
  env:
    OKTA_API_TOKEN: ${{ secrets.OKTA_API_TOKEN }}
  with:
    args: APP_PREFIX=trr APP_ORIGIN=web APP_TARGET=web URL_TARGET=https://%s.herokuapp.com/admin/okta OKTA_API_URL=https://login.therealreal.com/oauth2/v1/clients/0oa18r3unfqJuF9Y4357
- name: Remove From Okta Whitelist (ninja)
  uses: TheRealReal/heroku-review-envs/okta-whitelist-url-destroy@master
  if: "!contains( toJSON(github.event.pull_request.labels), 'okta-ninja')"
  env:
    GHA_USER_TOKEN: ${{ secrets.GHA_USER_TOKEN }}
    OKTA_API_TOKEN: ${{ secrets.OKTA_API_TOKEN_NINJA }}
  with:
    args: APP_PREFIX=trr APP_ORIGIN=web APP_TARGET=web URL_TARGET=https://%s.herokuapp.com/admin/okta
      OKTA_API_URL=https://login.therealreal.ninja/oauth2/v1/clients/0oalprgcvn7yBQMUU0h7
- name: Add To Okta Whitelist (ninja)
  uses: TheRealReal/heroku-review-envs/okta-whitelist-url-create@master
  if: "!contains( toJSON(github.event.pull_request.labels), 'okta-ninja')"
  env:
    OKTA_API_TOKEN: ${{ secrets.OKTA_API_TOKEN_NINJA }}
  with:
    args: APP_PREFIX=trr APP_ORIGIN=web APP_TARGET=web URL_TARGET=https://%s.herokuapp.com/admin/okta OKTA_API_URL=https://login.therealreal.ninja/oauth2/v1/clients/0oalprgcvn7yBQMUU0h7
- name: Remove From Okta Whitelist (com)
  uses: TheRealReal/heroku-review-envs/okta-whitelist-url-destroy@master
  if: "!contains( toJSON(github.event.pull_request.labels), 'okta-ninja')"
  env:
    GHA_USER_TOKEN: ${{ secrets.GHA_USER_TOKEN }}
    OKTA_API_TOKEN: ${{ secrets.OKTA_API_TOKEN }}
  with:
    args: APP_PREFIX=trr APP_ORIGIN=web APP_TARGET=web URL_TARGET=https://%s.herokuapp.com/admin/okta
      OKTA_API_URL=https://login.therealreal.com/oauth2/v1/clients/0oa18r3unfqJuF9Y4357
```

**Comment by dereksweet** [comment]
> Could we add a `README.md` like the other ones? I think what you've put in the PR description is almost the same.

Yes, absolutely. Meant to do that, sorry. It is there now.

File: heroku-config-var-set/heroku-config-var-set.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+args = {}
+for arg in sys.argv:
+    pair = arg.split('=')
+    if len(pair) > 1:
+        args[pair[0]] = '='.join(pair[1:])
+    else:
+        args[arg] = arg
+
+# print ("Found arguments: " + str( {k: v for k, v in args.items() if 'TOKEN' not in k and 'SECRET' not in k} ))
```
Comment (joepoon) [inline heroku-config-var-set/heroku-config-var-set.py]: nit - perhaps delete these commented lines or uncomment them for output during execution?

File: heroku-config-var-set/heroku-config-var-set.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    for pair in args['CONFIG_VARS'].split('|'):
+        (key, value) = pair.split('%')
+        config_vars[key] = value
+
+# print("Config Vars: %s" % ( config_vars ))
+
+app_prefix = args['APP_PREFIX']
+app_origin = args['APP_ORIGIN']
+app_target = args['APP_TARGET']
+# pr_num = args['PR_NUM']
```
Comment (joepoon) [inline heroku-config-var-set/heroku-config-var-set.py]: nit - remove this line? `# pr_num..`

File: heroku-config-var-set/heroku-config-var-set.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+# functions
+def get_app_name( svc_origin, svc_target, pr_num, prefix ):
+    if svc_origin == svc_target:
+        app_name = "%s-%s-pr-%s" % ( prefix, svc_origin, pr_num )
+    else:
+        app_name = "%s-%s-pr-%s-%s" % ( prefix, svc_origin, pr_num, svc_target )
+
+    return app_name
```
Comment (joepoon) [inline heroku-config-var-set/heroku-config-var-set.py]: More of a question, but I notice the other `get_app_name` methods return `name[:30]`.  Does this matter?


Example method from `heroku-app-destory.py` - looks almost identical (except if statement is flipped and `svc_name` vs `svc_target`)
```
def get_app_name( svc_origin, svc_name, pr_num, prefix ):
    if svc_origin != svc_name:
        # if related app, we append the app name
        name = "%s-%s-pr-%s-%s" % ( prefix, svc_origin, pr_num, svc_name )
    else:
        # if this is the originating app, just name it like vanilla review apps
        name = "%s-%s-pr-%s" % ( prefix, svc_origin, pr_num )
    # truncate to 30 chars for Heroku
    return name[:30]
```

File: heroku-config-var-set/heroku-config-var-set.py (hunk 1 above)
Comment (dereksweet) [reply heroku-config-var-set/heroku-config-var-set.py]: I kept them there for ease of development testing, but didn't want to output them like the other scripts do as it's setting config vars which may contain sensitive data. I can remove them if you want, but they sure are handy during dev.

File: heroku-config-var-set/heroku-config-var-set.py (hunk 2 above)
Comment (dereksweet) [reply heroku-config-var-set/heroku-config-var-set.py]: Again, kept this in here for ease of development testing as you don't have an actual PR to work with when running the commands locally. Can remove if you want.

Comment (joepoon) [reply heroku-config-var-set/heroku-config-var-set.py]: Oh I see what you mean... okay let's leave it there then.

File: heroku-config-var-set/heroku-config-var-set.py (hunk 3 above)
Comment (dereksweet) [reply heroku-config-var-set/heroku-config-var-set.py]: Yeah I'll put that back in there. I thought it was only important for creating the heroku app but of course when referencing the app you'd want to do the same. Good catch, thanks.

File: heroku-config-var-set/heroku-config-var-set.py (hunk 1 above)
Comment (joepoon) [reply heroku-config-var-set/heroku-config-var-set.py]: Sure, I'm fine with leaving it as is then (as it's consistent with the other scripts having this output line); you do have a point about sensitive data.

**Pull request merged**